Return a CType instead of a string for returns, beef up CType

bdhirsh opened the pull request:
Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55796 remove bridge API from codegen
* #55795 add _to_cpu() operator
* #55050 [EXPERIMENTAL, WIP] generate xla codegen in-tree
* #55334 split out NamedCType from CType. Remove direct string comparison from autograd codegen
* #55047 add C++ namespacing logic to ctypes
* **#55046 Return a CType instead of a string for returns, beef up CType**

Differential Revision: [D27708348](https://our.internmc.facebook.com/intern/diff/D27708348)

bdhirsh committed:
Return a CType instead of a string for returns, beef up CType

[ghstack-poisoned]

bdhirsh committed:
Update on "Return a CType instead of a string for returns, beef up CType"

Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.




[ghstack-poisoned]

bdhirsh committed:
Update on "Return a CType instead of a string for returns, beef up CType"

Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.




[ghstack-poisoned]

ezyang commented on tools/codegen/api/types.py:
```diff
@@ -72,7 +72,69 @@ def cpp_type(self, *, strip_ref: bool = False) -> str:
+    elem: 'CType'
+
+    def cpp_type(self, *, strip_ref: bool = False) -> str:
+        # Do not pass `strip_ref` recursively.
```
man `strip_ref` was a mistake haha. Since most of the time it doesn't go recursively maybe we can just make a small recursive pattern matching function that handles this.

ezyang commented on tools/codegen/api/types.py:
```diff
@@ -72,7 +72,69 @@ def cpp_type(self, *, strip_ref: bool = False) -> str:
+        # N.B. this isn't currently used anywhere: std::tuple is only used as a return type, which doesn't use names.
+        raise AssertionError("std::tuple isn't currently used as an argument anywhere, and doesn't require a name.")
+
+CType = Union[BaseCType, OptionalCType, ConstRefCType, MutRefCType, ListCType, ArrayRefCType, ArrayCType, VectorCType, TupleCType]
```
nit: probably wanna give each entry its own line soon

bdhirsh committed:
Update on "Return a CType instead of a string for returns, beef up CType"

Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.




[ghstack-poisoned]

bdhirsh committed:
Update on "Return a CType instead of a string for returns, beef up CType"

Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.




[ghstack-poisoned]

bdhirsh committed:
Update on "Return a CType instead of a string for returns, beef up CType"

Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.


Differential Revision: [D27708348](https://our.internmc.facebook.com/intern/diff/D27708348)

[ghstack-poisoned]

bdhirsh committed:
Update on "Return a CType instead of a string for returns, beef up CType"

Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.


Differential Revision: [D27708348](https://our.internmc.facebook.com/intern/diff/D27708348)

[ghstack-poisoned]

bhosmer commented on tools/codegen/api/types.py:
```diff
@@ -72,7 +72,69 @@ def cpp_type(self, *, strip_ref: bool = False) -> str:
+
+    @property
+    def name(self) -> ArgName:
+        return self.elem.name
```
Aha, turning `CType` into a little ADT is shining a brighter light on a structural problem with how "semantic information" (name, currently) is modeled here - it should really be associated with a complete type expression, not an expression's leaf nodes. 

The original composites skated on having exactly one leaf node, and most of the new ones here do too, but ofc the cracks are showing in `TupleCType` - also picture if we do a Map<K, V> type, etc.

It should be easy to refactor (and we should) but I wouldn't necessarily try to jam it into this PR, your call.

BTW, why not convert the originals (`OptionalCType` et al) as well?

bhosmer reviewed (approved):
Couple of observations inline, but nothing blocking - also, I haven't looked up the stack so not sure if they're superseded by later changes

bdhirsh committed:
Update on "Return a CType instead of a string for returns, beef up CType"

Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.


Differential Revision: [D27708348](https://our.internmc.facebook.com/intern/diff/D27708348)

[ghstack-poisoned]

bdhirsh committed:
Update on "Return a CType instead of a string for returns, beef up CType"

Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.


Differential Revision: [D27708348](https://our.internmc.facebook.com/intern/diff/D27708348)

[ghstack-poisoned]

bdhirsh committed:
Update on "Return a CType instead of a string for returns, beef up CType"

Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.


Differential Revision: [D27708348](https://our.internmc.facebook.com/intern/diff/D27708348)

[ghstack-poisoned]

bdhirsh committed:
Update on "Return a CType instead of a string for returns, beef up CType"

Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.


Differential Revision: [D27708348](https://our.internmc.facebook.com/intern/diff/D27708348)

[ghstack-poisoned]

bdhirsh committed:
Update on "Return a CType instead of a string for returns, beef up CType"

Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.


Differential Revision: [D27708348](https://our.internmc.facebook.com/intern/diff/D27708348)

[ghstack-poisoned]

bdhirsh committed:
Update on "Return a CType instead of a string for returns, beef up CType"

Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.


Differential Revision: [D27708348](https://our.internmc.facebook.com/intern/diff/D27708348)

[ghstack-poisoned]

bdhirsh committed:
Update on "Return a CType instead of a string for returns, beef up CType"

Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.


Differential Revision: [D27708348](https://our.internmc.facebook.com/intern/diff/D27708348)

[ghstack-poisoned]

bdhirsh committed:
Update on "Return a CType instead of a string for returns, beef up CType"

Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.


Differential Revision: [D27708348](https://our.internmc.facebook.com/intern/diff/D27708348)

[ghstack-poisoned]

bdhirsh committed:
Update on "Return a CType instead of a string for returns, beef up CType"

Updating `returns` in the codegen to return a CType instead of a raw string.

This has benefit of putting all stringifying logic through CType, which is useful in the followup PR when I add namespaces.

I also added new CTypes for other templated C++ types: array, vector and tuple. Mostly because it makes the namespacing logic in the next PR significantly easier. It also seems more natural to me that `BaseCType` shouldn't represent specializations of templated types.

There's a little bit of weirdness, types that are currently *only* used for returns, i.e. `TupleCType`. Returns aren't named, so I opted not to give it one- so we can add it in later if we discover that we need it.


Differential Revision: [D27708348](https://our.internmc.facebook.com/intern/diff/D27708348)

[ghstack-poisoned]
